### PR TITLE
feat(cli): support --quiet for lint command

### PR DESCRIPTION
**Pull request by thepassle**
implements `--quiet` as part of https://github.com/Boshen/oxc/issues/53

### CODE REVIEW

File: crates/oxc_cli/src/command.rs [hunk 1]
```diff
@@ -37,6 +37,11 @@ impl Command {
                     .help("File or Directory paths to scan. Directories are scanned recursively.")
                     .value_parser(ValueParser::path_buf()),
             )
+            .arg(
+              Arg::new("quiet")
```
Comment (thepassle) [inline crates/oxc_cli/src/command.rs:41]: Im not sure if I got the command quite right, it only works if I run it like:
```
cargo run --bin oxc_cli lint foo.js -- --quiet
```
(note the `--` before `--quiet`)

If I run it as:
```
cargo run --bin oxc_cli lint foo.js --quiet
```

I get:
```
⚡ cargo run --bin oxc_cli lint foo.js --quiet
    Finished dev [unoptimized + debuginfo] target(s) in 0.31s
     Running `target/debug/oxc_cli lint foo.js --quiet`
error: unexpected argument '--quiet' found

  note: to pass '--quiet' as a value, use '-- --quiet'

Usage: oxc lint <PATH> [quiet]

For more information, try '--help'.
```

File: crates/oxc_cli/src/options.rs [hunk 2]
```diff
@@ -0,0 +1,3 @@
+pub struct CliOptions {
```
Comment (thepassle) [inline crates/oxc_cli/src/options.rs:1]: in hindsight maybe its a bit overkill to put this in it's own file, although more CliOptions may be implemented over time, so I'm happy to hear your thoughts on this

**Comment by thepassle** [comment]
Gotta solve some Clippy warnings, but I'm gonna call it for the night, I'll take a look at those tomorrow 🙂

File: crates/oxc_cli/src/command.rs (hunk 1 above)
Comment (amitdahan) [reply crates/oxc_cli/src/command.rs:41]: What about `oxc_cli lint --quiet foo.js`?

File: crates/oxc_cli/src/options.rs (hunk 2 above)
Comment (Boshen) [reply crates/oxc_cli/src/options.rs:1]: This is good. We should be able to implement the `From` trait for clap commands.
So the `main` function will become smaller again.

File: crates/oxc_linter/src/rules/no_debugger.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@ use crate::{context::LintContext, rule::Rule};
 
 #[derive(Debug, Error, Diagnostic)]
 #[error("eslint(no-debugger): `debugger` statement is not allowed")]
-#[diagnostic()]
+#[diagnostic(severity(warning))]
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_debugger.rs:11]: Good catch!

File: crates/oxc_cli/src/main.rs [hunk 4]
```diff
@@ -21,7 +21,8 @@ fn main() -> CliRunResult {
                 return CliRunResult::PathNotFound { path: path.clone() };
             }
 
-            Cli::lint(path)
+            let quiet = matches.get_one::<String>("quiet").is_some();
+            Cli::lint(path, CliOptions { quiet })
```
Comment (Boshen) [inline crates/oxc_cli/src/main.rs:25]: We'll end up with 
```rust
let cli_options = CliOptions::From(clap matches);
let cli = Cli::new(cli_options);
cli.lint();
```

This should scale once we get even more top-level args and subargs.

File: crates/oxc_cli/src/lib.rs [hunk 5]
```diff
@@ -1,33 +1,41 @@
...
             .map(|path| {
                 let diagnostics = Self::lint_path(path);
-                for diagnostic in &diagnostics {
-                    println!("{diagnostic:?}");
-                }
+
+                diagnostics
+                    .iter()
+                    .filter(|diagnostic| {
+                        diagnostic.severity().unwrap() != Severity::Warning || !cli_options.quiet
```
[the remark this answers is not in the document]
Comment (thepassle) [reply crates/oxc_cli/src/lib.rs:33]: > Is quiet a filter or should it just disable the whole priting loop?

According to the eslint docs: 
image

**Comment by thepassle** [comment]
Hrm, feel like I've coded myself into a corner a little bit, @Boshen would you mind giving some pointers on the last commit? I was trying to implement `From` for the clap command, and then the compiler suggested I implement `TryFrom` instead because of the path potentially being faulty, but now I'm getting these warnings (see screenshot) and its making me feel like I took a wrong here:

image

File: crates/oxc_cli/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Boshen) [reply crates/oxc_cli/src/lib.rs]: lol I thought it's standard to not print anything. Do you mind adding a comment for this?

**Comment by Boshen** [comment]
> but now I'm getting these warnings (see screenshot) 

These are clippy commands, just follow the advice and add `#[must_use]` and `const`.

btw the `TryFrom` seems good to me.

**Comment by Boshen** [comment]
There's merge conflicts now, but you can do it!

File: crates/oxc_cli/src/options.rs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    type Error = &'a str;
+
+    fn try_from(matches: &ArgMatches) -> Result<Self, Self::Error> {
+        let path = matches.get_one::<PathBuf>("path").unwrap();
+        if path.canonicalize().is_err() {
+            return Err("err");
+        }
+
+        Ok(Self {
+            quiet: matches.get_one::<String>("quiet").is_some(),
```
Comment (jason89521) [inline crates/oxc_cli/src/options.rs:20]: Did you consider using `<bool>` instead of `<String>`?

Comment (thepassle) [reply crates/oxc_cli/src/options.rs:20]: I did yeah, but that didnt seem to work. I havent tried to make it its own command yet though, ill get to that today 🙂

File: crates/oxc_cli/src/command.rs (hunk 1 above)
Comment (thepassle) [reply crates/oxc_cli/src/command.rs:41]: Looks like `.action(clap::ArgAction::SetTrue)` should do the trick, will finish up the PR and then push my changes :)

File: crates/oxc_cli/src/options.rs (hunk 6 above)
Comment (thepassle) [reply crates/oxc_cli/src/options.rs:20]: Actually, in combination with `.action(clap::ArgAction::SetTrue)` `matches.get_one::<bool>` does work!

File: crates/oxc_cli/src/options.rs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+impl<'a> TryFrom<&'a ArgMatches> for CliOptions {
+    type Error = &'a str;
+
+    fn try_from(matches: &ArgMatches) -> Result<Self, Self::Error> {
+        let mut paths = vec![];
+
+        for path in matches.get_many::<PathBuf>("path").unwrap() {
+            let globbed = glob::glob(&path.to_string_lossy())
+                .expect("Failed to read glob pattern")
```
Comment (Boshen) [inline crates/oxc_cli/src/options.rs:18]: nitpick: this should not panic, return Err instead?

File: crates/oxc_cli/src/options.rs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                .map(|path| path.expect("Failed to read glob pattern"))
```
Comment (Boshen) [inline crates/oxc_cli/src/options.rs:19]: nitpick: this should not panic, return Err instead?

**Comment by thepassle** [comment]
Almost there, looks like `SetTrue` isnt quite behaving as I expected, its just always true now 🙃 Gotta do some things irl, will wrap this up today

File: crates/oxc_cli/src/lib.rs [hunk 9]
```diff
@@ -29,9 +45,17 @@ impl Cli {
...
             .map(|path| {
                 let diagnostics = Self::lint_path(path);
-                for diagnostic in &diagnostics {
-                    println!("{diagnostic:?}");
-                }
+
+                diagnostics
+                    .iter()
+                    .filter(|diagnostic| {
+                        diagnostic.severity().unwrap() != Severity::Warning
```
Comment (Boshen) [inline crates/oxc_cli/src/lib.rs:52]: Perhaps the comment 

> // This flag follows ESLint's --quiet behavior as documented here: https://eslint.org/docs/latest/use/command-line-interface#--quiet
> // Note that it does not disable ALL diagnostics, only Warning diagnostics

should be put here, this is confusing as hell lol

File: crates/oxc_cli/src/options.rs (hunk 7 above)
Comment (thepassle) [reply crates/oxc_cli/src/options.rs:18]: Yeah, I inherited this code from the glob PR, I can refactor it

File: crates/oxc_cli/src/lib.rs (hunk 9 above)
Comment (thepassle) [reply crates/oxc_cli/src/lib.rs:52]: Ah yeah I added it in `command.rs`, ill move it to here

File: crates/oxc_cli/src/options.rs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+                .collect::<Vec<_>>();
+
+            if globbed.is_empty() && path.canonicalize().is_err() {
+                return Err("Unable to find globbed files");
+            }
+            paths.extend(globbed);
+        }
+
+        Ok(Self { quiet: matches.get_one::<bool>("quiet").is_some(), paths })
```
Comment (Boshen) [inline crates/oxc_cli/src/options.rs:28]: try `get_flag`

**Comment by Boshen** [comment]
Squash, rebase, fix conflicts and we'll be good to go.

**Comment by thepassle** [comment]
@Boshen This should be good to go now :)

**Comment by thepassle** [comment]
My git-fu seems to be failing me, I cant seem to squash my commits 😐

**Comment by Boshen** [comment]
`git rebase -i HEAD~6`

**Comment by thepassle** [comment]
image

it seems like theres a bunch of commits in between, when I try to squash the commits I'm getting:
```
Cannot 'squash' without a previous commit

```

**Comment by Boshen** [comment]
What did you do lol. 

Do a reset --hard and start over?

Squash first then rebase.

**Comment by thepassle** [comment]
remade in https://github.com/Boshen/oxc/pull/72

**Pull request closed without merging**